src/Symfony/Component/TypeInfo/TypeResolver/TypeResolverInterface.php

```diff
@@ -0,0 +1,33 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\TypeInfo\TypeResolver;
+
+use Symfony\Component\TypeInfo\Exception\UnsupportedException;
+use Symfony\Component\TypeInfo\Type;
+use Symfony\Component\TypeInfo\TypeContext\TypeContext;
+
+/**
+ * Resolve a given subject to a Type.
+ *
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ * @author Baptiste Leduc <baptiste.leduc@gmail.com>
+ */
+interface TypeResolverInterface
+{
+    /**
+     * Try to resolve a {@see Type} on a $subject.
+     * If the resolver cannot resolve the type, it will throw a {@see UnsupportedException}.
+     *
+     * @throws UnsupportedException
+     */
+    public function resolve(mixed $subject, TypeContext $typeContext = null): Type;
```

Comment:
Some hint about what's expected to be passed as `$subject` would be useful. 
Maybe typehint on `\Reflector` also? (just wondering)

Reply:
I had set this as `mixed` first to avoid having BC when handling new stuff, but it should be safe to put it as `string|\Reflector` only. @mtarld, your opinion ?

Reply:
In my opinion, subject could be literally anything, that's why the mixed makes sense to me.

Maybe we can add some documentation to make it more understandable?

Reply:
Yes, that's the part of my proposal I'm sure about :) Also I'd suggest to narrow `$subect`'s type in each implementation via `@param` phpdoc since they're are all tight to a single type. Perhaps mention that null is not allowed in the interface also (if correct ofc)